[RadieonAjax] iP

RadieonAjax opened the pull request:
### Shrimp:

> Manage Tasks, Free Your Mind

Shrimp is a fast and lightweight task manager designed to keep you organized without getting in your way. Whether you need to keep track of tasks, manage deadlines, or get reminders, Shrimp has you covered.

## Why Shrimp?

- **Simple text-based interface**: No flashy UI—just type and go.
- **Fast and responsive**: Add and manage tasks without delay.
- **Perfect for Java learners**: Shrimp is an ideal tool for those looking to enhance their Java programming skills.

## How to Use:

1. **Download** the latest release version of Shrimp from [here](https://github.com/RadieonAjax/ip/releases).
2. **Double-click** to launch the application.
3. Start adding tasks by typing commands like `add`, `mark`, `delete`, etc.
4. Manage your tasks and let Shrimp take care of the details!

## What's Next?
- [x] **Add support for Deadline, Event and Todo** (https://github.com/RadieonAjax/ip/commit/658048f7d891096bbb656264484a1de87ffd6ec2)
- [x] **Add support for deletion** (https://github.com/RadieonAjax/ip/commit/0a09b7289d7ff530eeb83ed5ecb5b4586454ae56)
- [x] **Task persistence across sessions** (https://github.com/RadieonAjax/ip/commit/21151419eb56e0ccad72fcaf52cfe2f9fa317137)
- [x] **Graphical User Interface for easier access** (https://github.com/RadieonAjax/ip/commit/9406fc7b5462360946aa23a5e5d2a7fc1e183b45)
- [ ] **More to come...**

Stay tuned for more features to make your task management even easier!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

RadieonAjax committed:
Rename chatbot and implement initial greeting and exit functionality

- Renamed the chatbot to shrimp.
- Implemented a basic skeleton of the chatbot that greets the user upon start and exits after displaying a goodbye message.
- Replaced the source file name from Duke.java to Shrimp.java to reflect the new name.

RadieonAjax committed:
Add echo and exit functionality

- The chatbot now echo back any commands entered by the user.
- The loop terminates and the chatbot exits when the user types "bye".

RadieonAjax committed:
Add Task class

RadieonAjax committed:
Add TaskList class

RadieonAjax committed:
Add feature to store and list user commands

- Implemented the ability for the chatbot to store user-entered commands.
- Added a feature to display all stored commands in a numbered list when the user types "list".

RadieonAjax committed:
Add CommandParser class for enum usage

RadieonAjax committed:
Add support for completion check

RadieonAjax committed:
Add ability to mark and unmark tasks as done

- Implemented the ability to mark tasks as done and not done.
- Updated the task listing to reflect the current status of each task.
- Used an enum to handle the commands for better code structure and readability.

RadieonAjax committed:
Add Todo, Event and Deadline class

RadieonAjax committed:
Add support for new commands

RadieonAjax committed:
Add support for subclass polymorphism

RadieonAjax committed:
Add support for ToDos, Events, Deadlines

-Implemented support for ToDo, Deadline, and Event.
-Implemented polymorphism using an abstract Task class, with Todo, Deadline, and Event as subclasses.
-Enhanced task handling with inheritance and immutability by using a copy constructor to create new instances when marking tasks as done or not done.
-Refactored the list, mark, and unmark commands to work with the new task types and maintain immutability.
-Added a base case for commands to catch any errors.

RadieonAjax committed:
Add support for semi-automated testing

-Updated the test suite to compare Duke's actual output with expected results.

RadieonAjax committed:
Add support for ANSI

RadieonAjax committed:
Add ShrimpException for Exception Handling

-InvalidCommandException checks if the command is in the enum class.
-MissingArgumentException checks if the argument is present in the command given.
-InvalidArgumentException checks if the format which the argument given is valid.
-EmptyArrayException checks if the array is empty if the user issues a "list" command.

RadieonAjax committed:
Add support for array indexing issues

-Added ArrayIndexOutOfBoundException to check for indexing issues with the array.
-Renamed InvalidArgumentException to NumberFormatException in accordance with Java documentations.

RadieonAjax committed:
Add support for exception handling

-Implemented error handling for incorrect user inputs by introducing a custom exception class.
-Added specific error messages for cases such as missing task descriptions and unrecognized commands.
-Used exceptions to manage and throw errors, ensuring better code organization and improved user interaction.
-This update enhances the chatbot's robustness by providing informative feedback on errors and guiding the user to valid inputs.

RadieonAjax committed:
Add support for task deletion

RadieonAjax committed:
Add support for deletion exception handling

RadieonAjax committed:
Add support for command delete

RadieonAjax committed:
Add task deletion functionality to chatbot

-Implemented the delete command to allow users to remove tasks from their list.
-Updated TaskList class with a deleteTask method to handle task removal by index.
-Enhanced error handling for the delete command to manage invalid or out-of-range task numbers.
-Ensured that the deletion feature works across all task types (ToDo, Deadline, Event).

RadieonAjax committed:
Change program's file structure

RadieonAjax committed:
Update event types

RadieonAjax committed:
Update constructors and getters

RadieonAjax committed:
Add Storage class

RadieonAjax committed:
Implement task saving and loading functionality

- Introduced a Storage class to handle saving tasks to and loading tasks
 from a file.
- Tasks are automatically saved to ./data/shrimp.txt whenever the task
list changes.
- Tasks are loaded from the file when the chatbot starts up.
- Added error handling to manage cases where the file or directory does
not exist.
- Ensured that the file paths are specified in an OS-independent manner.

RadieonAjax committed:
Merge branch 'branch-Level-7'

RadieonAjax committed:
Add a constant PATTERN for datetime parsing

RadieonAjax committed:
Add InvalidDateTimeException

RadieonAjax committed:
Change output to fit the datetime datatype

RadieonAjax committed:
Add support for datetime datatype

RadieonAjax committed:
Add support for date and time parsing in tasks

- Implemented functionality to parse and handle dates and times for
tasks using `LocalDateTime`.
- Enhanced the parser to accept various date formats, including
single-digit days and months.
- Updated the `Deadline` and `Event` classes to store dates as
`LocalDateTime` objects instead of strings.
- Improved error handling to provide feedback when invalid date formats
are provided by the user.

RadieonAjax committed:
Merge branch 'branch-Level-8'

RadieonAjax committed:
Add Ui class to handle user interaction

RadieonAjax committed:
Add Command interface and the subclasses

RadieonAjax committed:
Extract classes for better code organization

-Refactored the codebase to follow object-oriented principles by
extracting related code into separate classes.
- Created the `Ui` class to handle interactions with the user, including
 displaying messages and reading user input.
- Implemented a command-based structure, with specific `Command` classes
 (e.g., `AddCommand`, `DeleteCommand`, etc.) inheriting from an abstract
  `Command` class.
- Simplified the main logic of the program, enabling better
maintainability and scalability.

RadieonAjax committed:
Merge branch 'branch-A-MoreOOP'

RadieonAjax committed:
Add Clear Command to clear list

RadieonAjax committed:
Updated automated testing

RadieonAjax committed:
Merge branch 'add-gradle-support'

RadieonAjax committed:
Add support for Gradle

RadieonAjax committed:
Update gradle configuration

RadieonAjax committed:
Update the main entry point for JAR file

RadieonAjax committed:
Optimize imports for files

RadieonAjax committed:
Add JUnit tests for core methods in Shrimp and Parser classes

- Added JUnit tests for the main logic in the Shrimp class, focusing on methods like `getTaskNumber()` and `getDateTime()`.
  - `getTaskNumber()` tests include scenarios for valid input, missing arguments, and invalid input formats.
  - `getDateTime()` tests cover valid date-time parsing and handling of invalid formats.

- Implemented tests for the Parser class to verify command parsing accuracy.
  - Tested command parsing for various command types, including `BYE`, `LIST`, `MARK`, `UNMARK`, `DEADLINE`, `EVENT`, `ADD`, `DELETE`, and `CLEAR`.
  - Added a test to ensure invalid commands are correctly identified as errors.

RadieonAjax committed:
Merge branch 'branch-A-JUnit'

RadieonAjax committed:
Add Javadocs for command types and exceptions

RadieonAjax committed:
Add Javadocs for task types

RadieonAjax committed:
Add Javadocs for Utiity classes and Shrimp

RadieonAjax committed:
Refactor coding style issues

RadieonAjax committed:
Add support for Find command

RadieonAjax committed:
Update TaskList for checking if empty

RadieonAjax committed:
Update exceptions for find command

RadieonAjax committed:
Integrate find command into chatbot

RadieonAjax committed:
Merge branch 'branch-A-JavaDoc'

RadieonAjax committed:
Resolve conflict for Merge

RadieonAjax committed:
Amend code structure issues

RadieonAjax committed:
Merge branch 'branch-Level-9'

RadieonAjax committed:
Resolve conflicts for Merge

RadieonAjax committed:
Merge commit 'f5f0b206e23d5227598e724fd7f9d5aedcd0d7d8'

miuna9890 commented on src/main/java/shrimp/task/Deadline.java:
```diff
@@ -0,0 +1,73 @@
+public class Deadline extends Task {
+    private final LocalDateTime by;
+
+    /**
+     * Constructs a {@code Deadline} task with the specified description, due date, and completion status.
+     *
+     * @param description The description of the task.
+     * @param by          The date and time by which the task should be completed.
+     * @param isDone      The completion status of the task.
+     */
```
Well done for commenting on all the parameters in ur header comment!

miuna9890 commented on src/main/java/shrimp/task/Event.java:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @return The start time of the event.
+     */
+    public LocalDateTime getFrom() {
```
getFrom seems a bit vague. For better understandability u could use getStartTime

miuna9890 commented on src/main/java/shrimp/task/Event.java:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @return The end time of the event.
+     */
+    public LocalDateTime getTo() {
```
getTo seems a bit vague. For better understandability u could use getEndTime

miuna9890 commented on src/main/java/shrimp/task/Task.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public abstract class Task {
+    private final String description; //description of the task
+    private final Boolean isDone; //boolean value for whether it's done or not
```
well done for remembering to use isDone for a boolean var!

miuna9890 commented on src/main/java/shrimp/task/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+     *
+     * @return The number of tasks in the list.
+     */
+    public int getCount() {
```
Could be better to name it as getNumOfTasks

miuna9890 commented on src/main/java/shrimp/utility/Storage.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        case "[E]" -> {
+            Event event = (Event) task;
+            return String.format("E | %d | %s | %s | %s", task.isDone() ? 1 : 0, task.getDescription(), event.getFrom(), event.getTo());
```
num of characters in this line seems too long. Good to make it 2 lines

jiaaaaen commented on src/main/java/shrimp/Shrimp.java:
```diff
@@ -0,0 +1,229 @@
+     *
+     * @param ui The {@code Ui} instance for interacting with the user.
+     */
+    static void chatBotRun(Ui ui) {
```
can be named runChatBot!

jiaaaaen commented on src/main/java/shrimp/Shrimp.java:
```diff
@@ -0,0 +1,229 @@
+                    } else if (taskList.getCount() == 0) {
+                        throw new ShrimpException.EmptyArrayException();
+                    }
+                    MarkCommand markUnmarkCommand = new MarkCommand(indexUnmark, false);
```
markUnmarkCommand is a little confusing

jiaaaaen commented on src/main/java/shrimp/Shrimp.java:
```diff
@@ -0,0 +1,229 @@
+     * @return The task number (zero-based index).
+     * @throws ShrimpException If the task number is missing or not a valid integer.
+     */
+    static int getTaskNumber(String userInput, Parser.CommandType type) throws ShrimpException {
```
type name could be more specific!

jiaaaaen commented on src/main/java/shrimp/utility/Parser.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Enum representing the various command types that can be parsed from user input.
+     */
+    public enum CommandType {
```
good use of enums!

jiaaaaen reviewed (commented):
looks good to me! organised and easy to understand :)

RadieonAjax committed:
Resolve checkstyle and import issues

RadieonAjax committed:
Merge branch 'branch-A-CheckStyle'

RadieonAjax committed:
Update gradle configuration for JavaFX

RadieonAjax committed:
Add main components for gui

RadieonAjax committed:
Modify run type to return string

RadieonAjax committed:
Modify methods to return string

RadieonAjax committed:
Update chatbot to separate duties

RadieonAjax committed:
Add images for chatbot

RadieonAjax committed:
Remove AnsiCode from text strings

RadieonAjax committed:
Update pre-check for errors

RadieonAjax committed:
Add title to JavaFX window

RadieonAjax committed:
Add styling to JavaFX window

RadieonAjax committed:
Merge branch 'branch-Level-10'

RadieonAjax committed:
README: Update to the latest feature

RadieonAjax committed:
README: Remove hyperlinks for features

RadieonAjax committed:
Update imports for Shrimp

RadieonAjax committed:
Add assertion for Command classes

RadieonAjax committed:
Enable assertation testing in build configuration

RadieonAjax committed:
Remove Ansi from strings

RadieonAjax committed:
Add assertation to utility classes

RadieonAjax committed:
Add assertions to document critical assumptions in code

- Implemented Java assert statements to document key assumptions in various parts of the code.
- Added assertions to validate conditions that should always hold, such as:
* Ensuring user input is not null or empty before processing commands.
* Validating task list boundaries when accessing tasks by index.
* Checking valid date and time formats for task deadlines and events.

RadieonAjax committed:
Refactor command package

RadieonAjax committed:
Refactor GUI package

RadieonAjax committed:
Refactor Task package

RadieonAjax committed:
Update method names after refactor

RadieonAjax committed:
Merge branch 'master'

RadieonAjax committed:
Refactor code to improve readability and maintainability

- Replaced ambiguous variable names with descriptive ones that make the code easier to understand without additional context.
- Extracted large, monolithic methods into smaller, well-defined methods in `Shrimp.java` to adhere to the Single Responsibility Principle (SRP), improving the structure and organization of the code.
- Ensured consistent naming conventions and streamlined flow for easier debugging and collaboration.

These changes enhance code readability and overall maintainability without altering functionality.

RadieonAjax committed:
Merge pull request #1 from RadieonAjax/branch-A-Assertions

Add assertations to Shrimp

RadieonAjax committed:
Merge branch 'master' into branch-A-CodeQuality

RadieonAjax committed:
Merge pull request #2 from RadieonAjax/branch-A-CodeQuality

Update code quality

RadieonAjax committed:
Merge branch 'master' into branch-Streams

RadieonAjax committed:
Update command parsing to use Map

RadieonAjax committed:
Merge pull request #3 from RadieonAjax/branch-A-Streams

Refactor command parsing logic using Streams and Map

RadieonAjax committed:
Create gradle.yml

RadieonAjax committed:
Update gradle.yml

RadieonAjax committed:
Update find command logic

Code has also been cleaned, with indentations in their correct positions.

RadieonAjax committed:
Update README.md

Updated the badge icons for CI and tag.

RadieonAjax committed:
Implement error message formatting for chatbot

RadieonAjax committed:
Enable circular profile picture

RadieonAjax committed:
Update layout for chatbot

RadieonAjax committed:
Improve GUI for better user experience

* Refined the overall look and feel of the chatbot interface.
* Customized label styles: created distinct visual formats for user input, bot replies, and error messages.
* Tweaked colors, padding, fonts, and alignments to make the conversation feel more natural and visually pleasing.
* Added support for circular profile pictures.
* Enabled responsive design: components: Shrimp now resize dynamically with window resizing, optimizing for various screen sizes.
* Removed unnecessary background space and streamlined the layout.
* Error highlighting to catch user attention when incorrect commands are entered.
* Improved overall UX by balancing visual design with functionality, ensuring the interface remains clean and easy to interact with.

RadieonAjax committed:
Merge pull request #5 from RadieonAjax/branch-A-BetterGui

Improve and polish GUI for better user experience

RadieonAjax committed:
Update boolean value to fit convention

RadieonAjax committed:
Merge pull request #6 from RadieonAjax/branch-A-Personality

- Gave the chatbot a unique personality by customizing its name 
to "Shrimp," enhancing the user interaction experience.
- Introduced custom responses for Shrimp, injecting humor and a 
conversational tone to align with its personality, providing a more 
engaging and friendly interaction.
- Refined the GUI with a distinct color palette and theme, reflecting 
Shrimp's playful and quirky nature.
- Updated fonts, icons, and layout to support Shrimp's 
personality, making the interface more visually appealing and consistent
 with the bot's character.
- Enhanced error messages and responses to be witty and less 
formal, aligning with Shrimp's quirky yet helpful persona.
- Tweaked background colors, padding, and alignments for a 
smoother, more visually pleasing conversation flow that resonates 
with Shrimp's cheerful vibe.
- Maintained a responsive design, ensuring all elements resize 
appropriately, while the playful visual cues remain intact.

RadieonAjax committed:
Enable trimming of inputs

RadieonAjax committed:
Add error handling for data integrity

RadieonAjax committed:
Add error handling for invalid characters

RadieonAjax committed:
Merge pull request #7 from RadieonAjax/branch-A-MoreErrorHandling

Enhanced error handling to address potential issues during usage:

- **Data Integrity Checks**: Added validation for event dates to 
ensure that the 'to' date is after the 'from' date. This prevents 
incorrect task scheduling.
- **Special Character Validation**: Implemented checks to handle 
special characters in task descriptions and commands, ensuring that 
only allowed characters are processed.

These improvements aim to make the chatbot more robust and reliable by 
catching common errors related to date handling and input validation.

RadieonAjax committed:
Add icon to chatbot

RadieonAjax committed:
Add Ui picture

RadieonAjax committed:
Update user guide

RadieonAjax committed:
Update README.md

RadieonAjax committed:
Merge branch 'master' into branch-A-JavaDoc

RadieonAjax committed:
Update javadocs for exceptions

RadieonAjax committed:
Add javadocs for gui

RadieonAjax committed:
Update javadocs for utility

RadieonAjax committed:
Update javadocs for shrimp

RadieonAjax committed:
Merge branch 'branch-A-JavaDoc'

RadieonAjax committed:
Merge branch 'master' into branch-A-CheckStyle

RadieonAjax committed:
Add checkstyle for project

RadieonAjax committed:
Update styling for command

RadieonAjax committed:
Update styling for gui

RadieonAjax committed:
Update styling for utility

RadieonAjax committed:
Update styling for JUnit tests

RadieonAjax committed:
Update styling for shrimp

RadieonAjax committed:
Update exception for event

RadieonAjax committed:
Merge branch 'branch-A-CheckStyle'

RadieonAjax committed:
Update gradle.yml

RadieonAjax committed:
Update gradle.yml

RadieonAjax committed:
Update README.md
